pkg/controller/bastion/actuator.go

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+
+	awssdk "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/gardener/gardener-extension-provider-aws/pkg/aws"
+	awsclient "github.com/gardener/gardener-extension-provider-aws/pkg/aws/client"
+	"github.com/gardener/gardener/extensions/pkg/controller/bastion"
+	"github.com/gardener/gardener/extensions/pkg/controller/common"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const (
+	sshPort                   = 22
+	instanceStateShuttingDown = 32
+	instanceStateTerminated   = 48
+)
+
+type actuator struct {
+	common.ClientContext
+
+	logger logr.Logger
+}
+
+func newActuator() bastion.Actuator {
+	return &actuator{
+		logger: logger,
+	}
+}
+
+func (a *actuator) getAWSClient(ctx context.Context, bastion *extensionsv1alpha1.Bastion, shoot *gardencorev1beta1.Shoot) (*awsclient.Client, error) {
+	secret := &corev1.Secret{}
+	key := types.NamespacedName{Name: v1beta1constants.SecretNameCloudProvider, Namespace: bastion.Namespace}
```

Comment:
FYI: There is `github.com/gardener/gardener/pkg/utils/kubernetes.Key` which might be more convenient

Reply:
I'm not sure the `Key()` function really offers anything of value here:

```
key := kubernetes.Key(v1beta1constants.SecretNameCloudProvider, bastion.Namespace)
// vs.
key := types.NamespacedName{Name: v1beta1constants.SecretNameCloudProvider, Namespace: bastion.Namespace}
```

Personally I would find the latter option to be more clear. I can change it if you insist though. ;)